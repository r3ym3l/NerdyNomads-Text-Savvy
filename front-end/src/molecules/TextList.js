/* eslint-disable */
import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";

import TextBox from "../atoms/TextBox/TextBox";
import TextBoxAdd from "../atoms/TextBox/TextBoxAdd";
import { updateWorkspace } from "../util/requests";
import "./TextList.css";


function TextList({ textList, workspaceId }) {
  const [ textItems, setTextItems ] = useState(null);

  const handleSubmit = (newTextItem) => {
    let newTextItems = [newTextItem, ...textItems];
    setTextItems(newTextItems);
    updateCurrentWorkspace(newTextItems);
  };

  const handleDelete = (id) => {
    // delete item in JSON based on id
    const newTextItems = textItems.filter(text => id !== text._id);
    setTextItems(newTextItems);
    updateCurrentWorkspace(newTextItems);
  };

  const updateCurrentWorkspace = async (items) => {
    let textIds = items.map( item => item._id );
    let updatedWorkspace = {
      texts: textIds,
      updateDate: Date.now()
    };
    await updateWorkspace(workspaceId, updatedWorkspace);
  }

  useEffect(() => {
    setTextItems(textList);
  }, [textList]);

  const renderList = () => {
    return textItems && textItems.map( (text) => {
      return <TextBox key={text._id} textItem={text} onDelete = {(id) => handleDelete(id)}/>;
    });
  };

  return (
    <div className= "text-list">
      {<TextBoxAdd workspaceId={workspaceId} showInput={false} onSubmit = {(newText) => handleSubmit(newText)}/>}
      {renderList()}
    </div>
  );
}

TextList.propTypes = {
  textList: PropTypes.array.isRequired,
  workspaceId: PropTypes.string.isRequired
};

export default TextList;