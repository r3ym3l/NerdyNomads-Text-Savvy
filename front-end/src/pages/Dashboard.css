.Dashboard {
    margin-left: 250px; /* size of sidebar */
    width: 100%; /* Fill the rest of the screen*/
}

.Dashboard-header-container {
    display: flex;
    justify-content: space-between;
    margin: 20px;
}

.Dashboard-title {
    font-weight: 600;
    font-size: 36px;
}

.Dashboard-reload-button {
    display: flex;
    justify-content: space-between;
    column-gap: 10px;
}