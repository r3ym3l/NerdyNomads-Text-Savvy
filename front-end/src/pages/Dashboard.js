import React, {useState, useEffect} from "react";
import PropTypes from "prop-types";

import TextList from "../molecules/TextList";
import "./Dashboard.css";
import { getWorkspaceInfo, getWorkspaceTexts } from "../util/requests";
import RefreshButton from "../atoms/RefreshButton";
import { ReloadIcon } from "../atoms/icons";

function Dashboard({ workspaceId }) {
  const [ textItems, setTextItems ] = useState(null);
  const [ renderedWorkspaceTitle, setRenderedWorkspaceTitle ] = useState("");

  const componentName = "Dashboard";

  const getCurrentWorkspace = async () => {
    if (workspaceId) {
      let workspaceResult = await getWorkspaceInfo(workspaceId);
      setRenderedWorkspaceTitle(workspaceResult?.data?.name);

      let textResult = await getWorkspaceTexts(workspaceId);
      setTextItems(textResult?.data);
    }
  };

  useEffect(() => {
    getCurrentWorkspace();
  }, [workspaceId]);

  const icon = (
    <div className={`${componentName}-reload-button`}>
      <div>Refresh</div>
      <ReloadIcon />
    </div>
  );

  return (
    <div className={`${componentName}`}>
      {/* Dashboard Header */}
      <div className={`${componentName}-header-container`}>
        <div className={`${componentName}-title`}>{renderedWorkspaceTitle}</div>
        <RefreshButton child={icon} onClick={() => location.reload()} />
      </div>
      {/* Dashboard Text */}
      { textItems &&
          <TextList textList={textItems} workspaceId={workspaceId}/>
      }
    </div>
  );
}

Dashboard.propTypes = {
  workspaceId: PropTypes.string.isRequired,
};

export default Dashboard;