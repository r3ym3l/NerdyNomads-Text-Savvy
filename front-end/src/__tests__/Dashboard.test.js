/* eslint-disable no-undef */
/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable react/display-name */
import React from "react";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { shallow, configure, mount } from "enzyme";
import { act } from "react-dom/test-utils";
import axios  from "axios";

import Dashboard from "../pages/Dashboard";
import TextList from "../molecules/TextList";

configure({ adapter: new Adapter() });  

jest.mock("axios");
jest.mock("../molecules/TextList", () => {
  return () => <div />;
});

const textItems = [
  { text: "Test1", source: "Test1" },
  { text: "Test1", source: "Test1" },
];

const whenStable = async () => {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
  });
};

test("Should match the snapshot.", () => { 
  let wrapper = shallow(<Dashboard workspaceId=""/>);
  expect(wrapper.html()).toMatchSnapshot();
});

test("TextList component should be displayed when the data has been received.", async () => { 
  axios.get
    .mockResolvedValueOnce({data: {name: ""}})
    .mockResolvedValueOnce({data: textItems});

  let realUseState = React.useState;
  let stubInitialState = textItems;
  jest.spyOn(React, "useState").mockImplementationOnce(() => {
    realUseState(stubInitialState);
  });

  let wrapper = mount(<Dashboard workspaceId="12345"/>);
  await whenStable();
  expect(axios.get).toBeCalledTimes(2);

  wrapper.update();
  expect(wrapper.containsMatchingElement(<TextList />)).toEqual(true);
});

test("Dashboard should show the workspace's name.", async () => { 
  let renderedWorkspaceTitle = "Test Workspace";
  axios.get.mockResolvedValueOnce({data: {name: renderedWorkspaceTitle}});

  let realUseState = React.useState;
  let stubInitialState = renderedWorkspaceTitle;
  jest.spyOn(React, "useState").mockImplementationOnce(() => {
    realUseState(stubInitialState);
  });

  let wrapper = mount(<Dashboard workspaceId="12345"/>);
  await whenStable();

  let titleDiv = wrapper.find(".Dashboard-title");
  expect(titleDiv.length).toEqual(1);  
  expect(titleDiv.first().text()).toEqual(renderedWorkspaceTitle);
});
