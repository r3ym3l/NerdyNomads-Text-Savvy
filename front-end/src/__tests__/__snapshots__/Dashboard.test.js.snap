// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match the snapshot. 1`] = `"<div class=\\"Dashboard\\"><div class=\\"Dashboard-header-container\\"><div class=\\"Dashboard-title\\"></div><div class=\\"Button RefreshButton\\" type=\\"submit\\"><div class=\\"Dashboard-reload-button\\"><div>Refresh</div><img alt=\\"reload\\" src=\\"/assets/images/reload_arrows.svg\\" class=\\"icon\\"/></div></div></div></div>"`;
